### PR TITLE
[azure] Optimize file status operations

### PR DESCRIPTION
**Pull request by jigold**
I think this change will help the number of operations we're making substantially. My Scala skills are not great, so I don't know if this is written correctly.

Basically, we were making a call to list the blobs recursively to test if the path was a directory which was streaming through the first 5000 records. I made the page size equal to 1 record as we don't care about all records. The next thing I did was to just get the blob properties rather than calling exists + get blob properties. So that will cut the number of HTTP calls by half for every blob. Lastly, listing items in a directory which is used for globbing was making a call to get the metadata for each file and then it was making the 3 API calls above to check whether it's a directory, whether it exists, and what the blob properties are. All of this information is in the original result from listing the blobs in the hierarchy so I just use that information directly.

### CODE REVIEW

File: build.yaml
Comment (danking) [inline build.yaml:2608]: note to ourselves to nix this before merging

File: hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
Comment (danking) [inline hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala:402]: holy yikes-a-molly! Great change, this is awful.

**Review by danking — CHANGES_REQUESTED** [review]
Wow this is a killer find. Excellent work. I think we should hold off on enabling QoB tests. Let's do that as a separate PR and maybe run it 10-20 times to convince ourselves its finally reliable first.

File: hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
Comment (danking) [inline hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala:153]: While you're changing build.yaml, also push this into the else branch.

File: hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala
Comment (danking) [inline hail/src/main/scala/is/hail/io/fs/AzureStorageFS.scala:402]: Ah, I think this is actually a bit of a bug. The name of a blobItem is like `/foo/bar/baz` but we need the URL which is `hail-az://account/container/foo/bar/baz`.

**Review by danking — DISMISSED** [review]

**Comment by jigold** [comment]
I'm putting the WIP tag to remind ourselves to turn off the Azure tests before merging.

**Review by danking — APPROVED** [review]
LGTM. Excellent change.

**Comment by jigold** [comment]
@danking I had to make some changes I'm not 100% confident in with stripping trailing "/".

**Review by danking — APPROVED** [review]
Yeah, this is right, the old code for directories did the same thing:
```scala
    val filename = dropTrailingSlash(url.toString)
    if (!isDir && !blobClient.exists()) {
      throw new FileNotFoundException(s"File not found: $filename")
    }

    if (isDir) {
      new BlobStorageFileStatus(path = filename, null, 0, isDir = true)
```

**Pull request merged**
